[MRG + 1] replace mutable default argument in _RidgeGCV

DonBeo opened the pull request:
This is a very small change to make me familiar with the repository and the git system.  The mutable default value is replaced with a numpy array. This code passes al the tests.

DonBeo committed:
replace mutable default argument in _RidgeGCV

DonBeo committed:
# This is a combination of 4 commits.
# The first commit's message is:
# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 3 commits.
# The first commit's message is:
# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 3 commits.
# The first commit's message is:
# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 3 commits.
# The first commit's message is:
# This is a combination of 2 commits.
# The first commit's message is:

documentation updated

use tuples instead of mutable  default values

solve conflict

up

conflict solved

conflict solved

remose white lines

delete white lines

use tuples instead of mutable  default values

use tuples instead of mutable  default values

redefinition of alphas is removed from init and added to fit method

conflict solved

use tuples instead of mutable  default values

documentation updated

use tuples instead of mutable  default values

redefinition of alphas is removed from init and added to fit method

conflict solved

use tuples instead of mutable  default values

redefinition of alphas is removed from init and added to fit method

conflict solved

documentation updated

remose white lines

delete white lines

update documentation

replace mutable default argument in _RidgeGCV

# This is the 2nd commit message:

redefinition of alphas is removed from init and added to fit method

# This is the 2nd commit message:

use tuples instead of mutable  default values

# This is the 3rd commit message:

conflict solved

# This is the 2nd commit message:

documentation updated

# This is the 2nd commit message:

remose white lines
# This is the 3rd commit message:

delete white lines

# This is the 2nd commit message:

redefinition of alphas is removed from init and added to fit method

# This is the 2nd commit message:

conflict solved

# This is the 3rd commit message:

use tuples instead of mutable  default values

# This is the 2nd commit message:

documentation updated

# This is the 2nd commit message:

redefinition of alphas is removed from init and added to fit method

# This is the 2nd commit message:

conflict solved

# This is the 2nd commit message:

redefinition of alphas is removed from init and added to fit method

# This is the 2nd commit message:

documentation updated

# This is the 3rd commit message:

remose white lines
# This is the 4th commit message:

delete white lines

DonBeo committed:
remove default mutable arguments

DonBeo committed:
replace mutable default argument in _RidgeGCV

DonBeo committed:
# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 3 commits.
# The first commit's message is:
# This is a combination of 4 commits.
# The first commit's message is:
# This is a combination of 18 commits.
# The first commit's message is:
replace mutable default argument in _RidgeGCV

# This is the 2nd commit message:

use tuples instead of mutable  default values

# This is the 3rd commit message:

use tuples instead of mutable  default values

# This is the 4th commit message:

redefinition of alphas is removed from init and added to fit method

# This is the 5th commit message:

use tuples instead of mutable  default values

# This is the 6th commit message:

joblib is restored to its original version. In ridge.py alphas is not casted to a numpy array anymore

# This is the 7th commit message:

collections library not required anymore and os it is removed

# This is the 8th commit message:

use tuples instead of mutable  default values

# This is the 9th commit message:

conflict solved

# This is the 10th commit message:

update documentation

# This is the 11th commit message:

documentation updated

# This is the 12th commit message:

use tuples instead of mutable  default values

# This is the 13th commit message:

solve conflict

# This is the 14th commit message:

up

# This is the 15th commit message:

conflict solved

# This is the 16th commit message:

conflict solved

# This is the 17th commit message:

remose white lines
# This is the 18th commit message:

delete white lines

# This is the 2nd commit message:

use tuples instead of mutable  default values

# This is the 3rd commit message:

use tuples instead of mutable  default values

# This is the 4th commit message:

redefinition of alphas is removed from init and added to fit method

# This is the 2nd commit message:

conflict solved

# This is the 3rd commit message:

use tuples instead of mutable  default values

# This is the 2nd commit message:

documentation updated

# This is the 2nd commit message:

use tuples instead of mutable  default values

# This is the 2nd commit message:

redefinition of alphas is removed from init and added to fit method

# This is the 2nd commit message:

conflict solved

# This is the 2nd commit message:

use tuples instead of mutable  default values

# This is the 2nd commit message:

redefinition of alphas is removed from init and added to fit method

DonBeo committed:
replace mutable default argument in _RidgeGCV

DonBeo committed:
replace mutable default argument in _RidgeGCV

DonBeo commented:
Sorry I did a mistake. The numpy array are mutable as well. Maybe we should stop this push. Anyway this should not influence the library at all.

amueller commented:
Well, you could use a tuple as you said. It would be good to add a test, though.

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

amueller commented:
Why did you close this?

DonBeo commented:
I was going to open a new one with the tuple instead of array. Is this not the correct procedure?

DonBeo commented:
Sorry I did not know that was possible to update the branch during the pull. I reopen the pull request. I hope that this does not lead to confusion. Regarding the test I need some help. I will ask on the mail list.

amueller commented:
You can ask here ;)
The pull request was already updated when you closed. The pull request will keep tracking your branch.
Actually, I don't think this can easily be tested. I'll try to add a test to the common_tests, but that might be a bit too tricky for this PR.

amueller commented:
I think this looks good to merge.
We'll need another review though. The travis failure is unrelated and I restarted the tests.

arjoly commented on sklearn/linear_model/ridge.py:
```diff
@@ -815,11 +815,11 @@ def identity_estimator():
                  score_func=None, loss_func=None, cv=None, gcv_mode=None,
                  store_cv_values=False):
-        self.alphas = alphas
+        self.alphas = np.asarray(alphas)
```
This line mutates the __init__ and we shouldn't.

amueller replied:
argh indeed. this needs to go into the ``fit`` method.

arjoly commented:
Whenever my remarks is addressed, +1!

DonBeo commented:
Sorry I need a clarification. 
Should we  never modify the init method? 
Should I change it?

Let me know!

amueller commented:
Yes, we should never modify anything in the ``__init__`` method, as parameters might get set via ``set_params`` and therefore bypass ``__init__``.
Please remove the conversion from ``__init__`` and create a local variable in ``fit`` that is set using the conversion.

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo commented:
I have replaced the self.alphas value in fit. I think this make the code more clear. Otherwise I can define a local variable alphas_loc = np.asarray(self.alphas) and replace all the occurencies of self.alphas with alphas_loc in the fit method. I do not know which approach is preffered.

amueller commented:
the second approach, though you can just call the variable ``alphas``.
The reason is that otherwise you overwrite the user input.

DonBeo committed:
redefinition of alphas is removed from init and added to fit method

DonBeo committed:
redefinition of alphas is removed from init and added to fit method

DonBeo committed:
redefinition of alphas is removed from init and added to fit method

DonBeo committed:
redefinition of alphas is removed from init and added to fit method

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo committed:
use tuples instead of mutable  default values

DonBeo commented:
in the linereader function it was quiet difficult to change the default value. I have been able to replace it but I am not happy of the quality of the solution.  Maybe it can be improved

arjoly commented on sklearn/linear_model/ridge.py:
```diff
@@ -847,8 +847,10 @@ def fit(self, X, y, sample_weight=None):
         -------
         self : Returns self.
         """
+        alphas = np.asarray(self.alphas)
```
Is it necessary?

DonBeo replied:
the tests are passed also without that line.  The problem is that the default argiment was a numpy array. If you want I can delete it and just use self.alphas in all the method.

arjoly replied:
> The problem is that the default argiment was a numpy array.

Have I missed something? It seems to be a list.

DonBeo replied:
check https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/ridge.py line 832.  The default argument for `init` was a numpy array.

arjoly replied:
I would go with the least number of lines to modify and go with the tuple.
Though, the doc says that it accepts (only) numpy array. http://scikit-learn.org/dev/modules/generated/sklearn.linear_model.RidgeClassifierCV.html#sklearn.linear_model.RidgeClassifierCV  From a user perpsective, I would expect array-like for alphas.

In that case, I would add a regression support and add support for array-like.

DonBeo replied:
ok then I think the last updated version should be ok.  I am using tuples as default value and we expect the users to use np.array.  I think that this should work  well. Unless I am not missing something. In that case explain me :-)

arjoly replied:
> In that case, I would add a regression support and add support for array-like.

I mean. I would add a test for this. By the way, we should update the documentation.

DonBeo replied:
I am afraid that you have to help me with test. I am not sure of what to test and how to do it.

arjoly replied:
For instance, you could test that passing a tuple of alpha, a numpy array of alpha or a list of alpha give similar results.

DonBeo replied:
ok I have added a test. Let me know if you think that this is fine.

arjoly commented on sklearn/externals/joblib/format_stack.py:
```diff
@@ -264,7 +264,13 @@ def tokeneater(token_type, token, start, end, line):
         tokeneater.name_cont = False
 
-        def linereader(file=file, lnum=[lnum], getline=linecache.getline):
+        def linereader(file=file, lnum_loc=lnum, getline=linecache.getline):
```
This patch should be done upstream in the joblib repository.

arjoly replied:
It's also better if the pr stays focus.

DonBeo replied:
What should I do? Do I restore the original version?

arjoly replied:
+1 for the original version.

DonBeo replied:
ok I have restored the original version of this.

DonBeo committed:
joblib is restored to its original version. In ridge.py alphas is not casted to a numpy array anymore

DonBeo committed:
joblib is restored to its original version. In ridge.py alphas is not casted to a numpy array anymore

DonBeo committed:
joblib is restored to its original version. In ridge.py alphas is not casted to a numpy array anymore

DonBeo committed:
joblib is restored to its original version. In ridge.py alphas is not casted to a numpy array anymore

DonBeo committed:
collections library not required anymore and os it is removed

DonBeo committed:
collections library not required anymore and os it is removed

DonBeo committed:
collections library not required anymore and os it is removed

DonBeo committed:
collections library not required anymore and os it is removed

DonBeo committed:
test added to verify that both tuples and np.array can be used as alpha value in ridgeCV

DonBeo commented:
I think that this conflict can be caused by the test.  My understanding is that to solve it I have to write the original sklearn repository.  I think that only a moderator can do that.

arjoly commented:
With some git-fu, you can synchronise your branch. 
Try to rebase on top of master as explained in http://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html.

DonBeo committed:
conflict solved

DonBeo committed:
conflict solved

DonBeo committed:
solved conflict

DonBeo commented:
I would suggest to add this link  http://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html. to the contributing documentation. It was very helpfull for me. People like me who  have only a basic understanding of git often do not collaborate because don't know exactly where to start with git. 

By the way hopefully the merge is completed :-)

arjoly commented:
With a small update to the class doc ([here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/ridge.py#L903) and [there](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/ridge.py#L987)), you have my +1.

DonBeo committed:
update documentation

DonBeo committed:
update documentation

DonBeo committed:
update documentation

DonBeo commented:
The docs is updated so hopefully we can do the merge. 

By the way I am looking for a mentor for the GSoC.  I think that now that I have an understanding of how git works I will not need much supervision. Here there is my proposal https://docs.google.com/document/d/1gCHUKsfvii1sUQW-4E4dpbpWkmTPAg6WpLUcWbu4vk0/edit?usp=sharing  .

Let me know!!

amueller commented on sklearn/linear_model/ridge.py:
```diff
@@ -900,7 +901,7 @@ class RidgeCV(_BaseRidgeCV, RegressorMixin):
     Parameters
     ----------
-    alphas : numpy array of shape [n_alphas]
+    alphas : numpy array or tuple of length [n_alphas]
```
It think it should be array-like.

amueller commented on sklearn/linear_model/ridge.py:
```diff
@@ -984,7 +985,7 @@ class RidgeClassifierCV(LinearClassifierMixin, _BaseRidgeCV):
     Parameters
     ----------
-    alphas : numpy array of shape [n_alphas]
+    alphas : numpy array or tuple of length [n_alphas]
```
same here.

DonBeo replied:
Sorry let's try to find a common agreement. The original was 
`- alphas : numpy array of shape [n_alphas]`  
Can I replace it with: `array-like of shape [n_alphas]` ?

amueller replied:
exactly.

amueller commented:
There seemed to have been some issue with tracking the changes, it seems.
If you want, you can try to fix it by rebasing onto current master.

DonBeo committed:
documentation updated

DonBeo committed:
documentation updated

DonBeo committed:
documentation updated

DonBeo commented:
ok hopefully everything is fine now.  How did you know that there where some issues with tracking? How can I check that?

amueller commented:
The "commits" and a "files changed" tabs above show unrelated commits and changes.

DonBeo commented:
The code should be fine now. Let me know if it can be merged or if I need to do something else

amueller commented:
Still showing the same commits and changes. Can you
```
git checkout master
git pull upstream master
git checkout my-feature
git rebase master
```
and fix possible conflicts.
Then do 
``git rebase -i master`` and squash them.

amueller commented:
Or maybe just leave it and I'll do that once we get another review.

DonBeo committed:
conflict solved

DonBeo committed:
solve conflict

DonBeo committed:
solve conflict

DonBeo committed:
up

DonBeo committed:
up

DonBeo committed:
conflict solved

DonBeo committed:
conflict solved

DonBeo committed:
conflict solved

DonBeo committed:
conflict solved

DonBeo committed:
merged

ogrisel commented on sklearn/externals/joblib/format_stack.py:
```diff
@@ -264,7 +263,11 @@ def tokeneater(token_type, token, start, end, line):
 
+            
+           
+
```
Please do not introduce unnessary blank lines. Please run the [pep8](https://pypi.python.org/pypi/pep8) check tool when not sure.

ogrisel replied:
actually please ignore this apparently this commit was not supposed to be included in the PR.

ogrisel commented on sklearn/externals/joblib/format_stack.py:
```diff
@@ -264,7 +263,11 @@ def tokeneater(token_type, token, start, end, line):
+           
+
         def linereader(file=file, lnum=[lnum], getline=linecache.getline):
+            
```
Same here. This new blank line is useless.

ogrisel commented:
This branch has a mix of commits from master and your own. Please clean them up. Also never merge anything into you branch if you plan to rebase it on master.

ogrisel commented:
Also fixes to joblib should happen in the upstream joblib/joblib repo otherwise they will be lost at the next code sync.

DonBeo committed:
remose white lines

DonBeo committed:
remose white lines

DonBeo committed:
delete white lines

DonBeo committed:
delete white lines

DonBeo commented:
Sorry this was my first time with github and I did some confusion. 
All the changes in joblib folder have been removed and the file inside should be equivalent to the original ones.  

Regarding the commit I need to read better the documentation because I do not know how to clean.

amueller committed:
FIX LDA(solver="lsqr"): make sure the right error is raised on transform.

amueller commented:
did you see my instructions from two days ago?

Btw @ogrisel wasn't there a PR to include a squashing guide in the dev docs? I can't find it any more :-/

DonBeo committed:
rebase

DonBeo committed:
removed default mutable argument

DonBeo commented:
Hi, 
I did rebase again. Hopefully now it should be fine. 
My understanding is that with the command `git rebase -i master` I can delete some of the commits but at the moment all the commits are related to changes that I did. 

Let me know if this is fine.

Sorry for the confusion I am starting to understand git and the next pull will hopefully be less painful

amueller commented:
when you do rebase -i, you can mark all but the first commit as ``squash`` so they will be resolved to a single commit.

DonBeo committed:
# This is a combination of 4 commits.
# The first commit's message is:
conflict solved

# This is the 2nd commit message:

documentation updated

# This is the 3rd commit message:

remose white lines
# This is the 4th commit message:

delete white lines

DonBeo committed:
update documentation

DonBeo committed:
conflict solved

DonBeo commented:
I did rebase again. The number of commits now is lower. My understanding is that I can not have a single commit for certain addition.

I hope that this if fine now.

raghavrv committed:
MAINT docstring --> comments to prevent nose from using doc in verbose mode

raghavrv committed:
MAINT use yield for cleaner output in verbose mode

ogrisel committed:
Merge pull request #4432 from ragv/travis_ignore_docstring

[MRG + 1] Prevent nose from using docstring to name the tests in results

amueller committed:
COSMIT missing newlines in metrics test

amueller commented:
Can you please look at the commit count shown a the top? It shows 97 commits.
Also, you can do anything within a single commit (apart from a merge, which you should not have).

amueller committed:
Merge pull request #4436 from ogrisel/rebased-pr-3747

[MRG+2] FIX make StandardScaler & scale more numerically stable

DonBeo committed:
remove default mutable arguments

ogrisel committed:
Merge pull request #4427 from amueller/lda_lsqr_predict

[MRG+1] FIX LDA(solver="lsqr"): make sure the right error is raised on transform

DonBeo committed:
remove default mutable values

DonBeo committed:
conflict solved

DonBeo committed:
conflict solved

amueller commented:
sorry, I'd be happy to help you, but I just missed you on IRC. you can ping me as amueller, then something will bounce on my desktop ;)

DonBeo committed:
solved conflict

TomDLT commented:
#4713 merged
This PR can be closed
